watchtower/wtclient/session_queue.go

```diff
@@ -151,41 +152,74 @@ func newSessionQueue(cfg *sessionQueueConfig,
 // backups.
 func (q *sessionQueue) Start() {
 	q.started.Do(func() {
+		q.wg.Add(1)
 		go q.sessionManager()
 	})
 }
 
 // Stop idempotently stops the sessionQueue by initiating a clean shutdown that
 // will clear all pending tasks in the queue before returning to the caller.
-func (q *sessionQueue) Stop() {
+func (q *sessionQueue) Stop() error {
+	var returnErr error
+
 	q.stopped.Do(func() {
 		q.log.Debugf("SessionQueue(%s) stopping ...", q.ID())
 
+		shutdown := make(chan struct{})
+		go func() {
+			for {
+				select {
+				case <-time.After(time.Millisecond):
+					q.queueCond.Signal()
+				case <-shutdown:
+					return
+				}
+			}
+		}()
+
 		close(q.quit)
-		q.signalUntilShutdown()
+		q.wg.Wait()
+		close(shutdown)
 
-		// Skip log if we also force quit.
-		select {
-		case <-q.forceQuit:
+		// Now, for any task in the pending queue that we have not yet
+		// created a CommittedUpdate for, re-add the task the main
+		// task pipeline.
+		updates, err := q.cfg.DB.FetchSessionCommittedUpdates(q.ID())
+		if err != nil {
+			returnErr = err
 			return
-		default:
 		}
 
-		q.log.Debugf("SessionQueue(%s) stopped", q.ID())
-	})
-}
+		unAckedUpdates := make(map[wtdb.BackupID]bool)
+		for _, update := range updates {
+			unAckedUpdates[update.BackupID] = true
+		}
 
-// ForceQuit idempotently aborts any clean shutdown in progress and returns to
-// he caller after all lingering goroutines have spun down.
-func (q *sessionQueue) ForceQuit() {
-	q.forced.Do(func() {
-		q.log.Infof("SessionQueue(%s) force quitting...", q.ID())
+		// of each session onto the task pipeline.
+		q.queueCond.L.Lock()
```

Comment:
Hmm, do we ever unlock this? Maybe this is the problem and you can ignore my previous comment.

Reply:
I added a `defer q.queueCond.L.Unlock()` here and ran the tests a couple of times. The race deadlock didn't happen anymore, but 1 out of 4 runs failed with:
```
--- FAIL: TestClient (0.00s)
    --- FAIL: TestClient/committed_update_restart (2.04s)
        client_test.go:1234: 
                Error Trace:    /home/guggero/projects/go/src/github.com/lightningnetwork/lnd/watchtower/wtclient/client_test.go:1011
                                                        /home/guggero/projects/go/src/github.com/lightningnetwork/lnd/watchtower/wtclient/client_test.go:1234
                                                        /home/guggero/projects/go/src/github.com/lightningnetwork/lnd/watchtower/wtclient/client_test.go:2456
                Error:          Should be true
                Test:           TestClient/committed_update_restart
                Messages:       breach hints not received, only got 2/3
    --- FAIL: TestClient/can_remove_tower_with_un-acked_updates_(with_restart) (6.23s)
        client_test.go:2441: Received 2/3
        client_test.go:2441: Received 2/3
        client_test.go:2441: Received 2/3
        client_test.go:2441: Received 2/3
        client_test.go:2441: 
                Error Trace:    /home/guggero/projects/go/src/github.com/lightningnetwork/lnd/watchtower/wtclient/client_test.go:1011
                                                        /home/guggero/projects/go/src/github.com/lightningnetwork/lnd/watchtower/wtclient/client_test.go:2441
                                                        /home/guggero/projects/go/src/github.com/lightningnetwork/lnd/watchtower/wtclient/client_test.go:2456
                Error:          Should be true
                Test:           TestClient/can_remove_tower_with_un-acked_updates_(with_restart)
                Messages:       breach hints not received, only got 2/3
FAIL
FAIL    github.com/lightningnetwork/lnd/watchtower/wtclient     19.992s
```

So something's flaky either way.